File: Data Structures & Algorithms/search-2d-matrix/submission-1.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def binarySearch(row):
            l, r = 0, len(row)
            while l <= r:
                mid = l + (r-l) // 2
                if row[mid] > target:
                    r = mid -1
                elif row[mid] < target:
                    l = mid + 1
                else:
                    return True
            return False
        R, C = len(matrix), len(matrix[0])
        t, b = 0, R-1
        while t <= b:
            mid = t + (b-t) // 2
            if matrix[mid][0] > target:
                b = mid - 1
            elif matrix[mid][-1] < target:
                t = mid + 1
            else:
                break
        if not (t <= b):
            return False
        row = (t+b) // 2
        return binarySearch(matrix[row])

            

File: Data Structures & Algorithms/set-zeroes-in-matrix/submission-1.cpp
class Solution {
public:
  void setZeroes(vector<vector<int>>& matrix) {
    int n = matrix.size();
    int m = matrix[0].size();
    bool rowZero = false;
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (matrix[i][j] == 0) {
          matrix[0][j] = 0;
          if (i > 0) {
            matrix[i][0] = 0;
          } else {
            rowZero = true;
          }
        }
      }
    }
    for (int i = 1; i < n; i++) {
      for (int j = 1; j < m; j++) {
        if (matrix[i][0] == 0 || matrix[0][j] == 0) {
          matrix[i][j] = 0;
        }
      }
    }
     if (matrix[0][0] == 0) {
            for (int r = 0; r < n; r++) {
                matrix[r][0] = 0;
            }
        }
    if (rowZero) {
      for (int j = 0; j < m; j++) {
        matrix[0][j] = 0;
      }
    }
  }



};
